File: src/subsystem/Charts/apis/Yahoo.ts
/**
 * Yahoo Finance Data Source Module
 * Provides stock market data from Yahoo Finance API
 */
export interface YahooStockItem {
    symbol: string;
    name: string;
    currentPrice: number;
    changePercent: number;
    changeAmount: number;
    volume: number;
    marketCap?: number;
    sector?: string;
    exchange?: string;
    lastUpdated?: number;
    open?: number;
    high?: number;
    low?: number;
    previousClose?: number;
}
export interface YahooChartResult {
    meta: {
        currency: string;
        symbol: string;
        exchangeName: string;
        fullExchangeName: string;
        instrumentType: string;
        regularMarketPrice: number;
        regularMarketChange?: number;
        regularMarketChangePercent?: number;
        regularMarketTime: number;
        regularMarketVolume: number;
        regularMarketDayHigh: number;
        regularMarketDayLow: number;
        regularMarketPreviousClose?: number;
        chartPreviousClose?: number;
        longName?: string;
        shortName?: string;
        priceHint: number;
        dataGranularity: string;
        range: string;
    };
    timestamp: number[];
    indicators: {
        quote: Array<{
            open: (number | null)[];
            high: (number | null)[];
            low: (number | null)[];
            close: (number | null)[];
            volume: (number | null)[];
        }>;
    };
}
export interface YahooFinanceResponse {
    chart: {
        result: YahooChartResult[];
        error: any;
    };
}
// Popular US stocks for default tracking
export const POPULAR_STOCKS = [
    "AAPL", "MSFT", "GOOGL", "AMZN", "META",
    "TSLA", "NVDA", "BRK-B", "JPM", "V",
    "JNJ", "WMT", "UNH", "PG", "MA",
    "HD", "CVX", "BAC", "XOM", "PFE"
];
/**
 * Fetch stock data from Yahoo Finance
 * @param symbol - Stock symbol (e.g., "AAPL")
 * @returns Stock data or null if failed
 */
export const fetchStockFromYahoo = async (symbol: string): Promise<YahooStockItem | null> => {
    try {
        const url = `https://query1.finance.yahoo.com/v8/finance/chart/${symbol}?range=1d&interval=5m`;
        const response = await fetch(url, {
            headers: {
                Accept: "application/json",
                "User-Agent": "Mozilla/5.0",
            },
        });
        if (!response.ok) {
            throw new Error(`HTTP ${response.status}`);
        }
        const data: YahooFinanceResponse = await response.json();
        if (data.chart.error || !data.chart.result || data.chart.result.length === 0) {
            return null;
        }
        const result = data.chart.result[0];
        const meta = result.meta;
        const indicators = result.indicators;
        const quote = indicators.quote[0];
        const currentPrice = meta.regularMarketPrice;
        const previousClose = meta.regularMarketPreviousClose || meta.chartPreviousClose || currentPrice * 0.99;
        const changeAmount = currentPrice - previousClose;
        const changePercent = (changeAmount / previousClose) * 100;
        return {
            symbol: meta.symbol,
            name: meta.longName || meta.shortName || symbol,
            currentPrice,
            changePercent,
            changeAmount,
            volume: meta.regularMarketVolume,
            marketCap: undefined,
            sector: undefined,
            exchange: meta.fullExchangeName,
            lastUpdated: meta.regularMarketTime * 1000,
            open: quote.open[0] || undefined,
            high: meta.regularMarketDayHigh,
            low: meta.regularMarketDayLow,
            previousClose: previousClose,
        };
    } catch (err) {
        console.warn(`[Yahoo] Failed to fetch ${symbol}:`, err);
        return null;
    }
};
/**
 * Fetch multiple stocks data from Yahoo Finance
 * @param symbols - Array of stock symbols
 * @param onProgress - Optional progress callback
 * @returns Array of successful stock data
 */
export const fetchStocksBatch = async (
    symbols: string[],
    onProgress?: (current: number, total: number) => void
): Promise<YahooStockItem[]> => {
    const results: YahooStockItem[] = [];
    const total = symbols.length;
    for (let i = 0; i < total; i++) {
        const symbol = symbols[i];
        try {
            const stockData = await fetchStockFromYahoo(symbol);
            if (stockData) {
                results.push(stockData);
            }
            // Rate limiting to avoid being blocked
            if (i < total - 1) {
                await new Promise((resolve) => setTimeout(resolve, 300));
            }
        } catch (err) {
            console.warn(`[Yahoo] Failed to fetch ${symbol}:`, err);
        }
        if (onProgress) {
            onProgress(i + 1, total);
        }
    }
    return results;
};
/**
 * Generate fallback stock data when API is unavailable
 */
export const createFallbackStockData = (): YahooStockItem[] => {
    const stockTemplates = [
        { symbol: "AAPL", name: "Apple Inc.", basePrice: 185, sector: "Technology" },
        { symbol: "MSFT", name: "Microsoft Corp.", basePrice: 415, sector: "Technology" },
        { symbol: "GOOGL", name: "Alphabet Inc.", basePrice: 153, sector: "Technology" },
        { symbol: "AMZN", name: "Amazon.com Inc.", basePrice: 178, sector: "Consumer" },
        { symbol: "TSLA", name: "Tesla Inc.", basePrice: 245, sector: "Automotive" },
        { symbol: "NVDA", name: "NVIDIA Corp.", basePrice: 950, sector: "Technology" },
        { symbol: "META", name: "Meta Platforms", basePrice: 485, sector: "Technology" },
        { symbol: "JPM", name: "JPMorgan Chase", basePrice: 189, sector: "Financial" },
        { symbol: "V", name: "Visa Inc.", basePrice: 279, sector: "Financial" },
        { symbol: "JNJ", name: "Johnson & Johnson", basePrice: 161, sector: "Healthcare" },
    ];
    return stockTemplates.map((stock) => {
        const changePercent = (Math.random() - 0.5) * 5;
        const currentPrice = stock.basePrice * (1 + changePercent / 100);
        const changeAmount = currentPrice - stock.basePrice;
        const volume = Math.random() * 1e7 + 1e6;
        return {
            symbol: stock.symbol,
            name: stock.name,
            currentPrice,
            changePercent,
            changeAmount,
            volume,
            marketCap: stock.basePrice * (Math.random() * 5e6 + 1e6),
            sector: stock.sector,
            exchange: "NASDAQ",
            lastUpdated: Date.now(),
        };
    });
};

File: src-tauri/src/subsystem/finance/yahoo.rs
use crate::commons::https::HttpClient;
use serde::{Deserialize, Serialize};
use tauri::command;
/// Yahoo Finance stock data structure
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct YahooStockData {
    pub symbol: String,
    pub name: String,
    pub current_price: f64,
    pub change_percent: f64,
    pub change_amount: f64,
    pub volume: u64,
    pub market_cap: Option<f64>,
    pub sector: Option<String>,
    pub exchange: Option<String>,
    pub open: Option<f64>,
    pub high: Option<f64>,
    pub low: Option<f64>,
    pub previous_close: Option<f64>,
}
/// Batch response
#[derive(Debug, Serialize, Deserialize)]
pub struct YahooBatchResponse {
    pub success: bool,
    pub message: String,
    pub total: usize,
    pub data: Vec<YahooStockData>,
}
/// Yahoo Finance fetcher
pub struct YahooFetcher {
    http: HttpClient,
}
impl YahooFetcher {
    pub fn new() -> Self {
        Self { http: HttpClient::new() }
    }
    /// Fetch single stock from Yahoo Finance
    pub async fn fetch_stock(&self, symbol: &str) -> Result<YahooStockData, String> {
        let url = format!("https://query1.finance.yahoo.com/v8/finance/chart/{}?range=1d&interval=5m", symbol);
        let response = self.http.fetch_json(&url, None).await.map_err(|e| format!("HTTP error: {}", e))?;
        // Parse response
        let chart = response.get("chart").ok_or("Missing chart field")?;
        let error = chart.get("error");
        if error.is_some() && !error.as_ref().unwrap().is_null() {
            return Err("Yahoo API returned error".to_string());
        }
        let result = chart.get("result").and_then(|v| v.as_array()).and_then(|arr| arr.first()).ok_or("No result data")?;
        let meta = result.get("meta").ok_or("Missing meta")?;
        let indicators = result.get("indicators").ok_or("Missing indicators")?;
        let quote = indicators.get("quote").and_then(|v| v.as_array()).and_then(|arr| arr.first()).ok_or("Missing quote data")?;
        let current_price = meta.get("regularMarketPrice").and_then(|v| v.as_f64()).unwrap_or(0.0);
        let previous_close = meta
            .get("regularMarketPreviousClose")
            .and_then(|v| v.as_f64())
            .or_else(|| meta.get("chartPreviousClose").and_then(|v| v.as_f64()))
            .unwrap_or(current_price * 0.99);
        let change_amount = current_price - previous_close;
        let change_percent = if previous_close > 0.0 { (change_amount / previous_close) * 100.0 } else { 0.0 };
        let symbol_name = meta.get("symbol").and_then(|v| v.as_str()).unwrap_or(symbol).to_string();
        let long_name = meta
            .get("longName")
            .and_then(|v| v.as_str())
            .or_else(|| meta.get("shortName").and_then(|v| v.as_str()))
            .unwrap_or(&symbol_name)
            .to_string();
        let open = quote.get("open").and_then(|v| v.as_array()).and_then(|arr| arr.first()).and_then(|v| v.as_f64());
        let high = meta.get("regularMarketDayHigh").and_then(|v| v.as_f64());
        let low = meta.get("regularMarketDayLow").and_then(|v| v.as_f64());
        let volume = meta.get("regularMarketVolume").and_then(|v| v.as_u64()).unwrap_or(0);
        let exchange = meta.get("fullExchangeName").and_then(|v| v.as_str()).map(|s| s.to_string());
        Ok(YahooStockData {
            symbol: symbol_name,
            name: long_name,
            current_price,
            change_percent,
            change_amount,
            volume,
            market_cap: None,
            sector: None,
            exchange,
            open,
            high,
            low,
            previous_close: Some(previous_close),
        })
    }
    /// Fetch multiple stocks from Yahoo Finance
    pub async fn fetch_stocks_batch(&self, symbols: &[&str]) -> Vec<YahooStockData> {
        let mut results = Vec::new();
        for (i, &symbol) in symbols.iter().enumerate() {
            match self.fetch_stock(symbol).await {
                Ok(data) => results.push(data),
                Err(e) => eprintln!("[Yahoo] Failed to fetch {}: {}", symbol, e),
            }
            // Rate limiting to avoid being blocked
            if i < symbols.len() - 1 {
                tokio::time::sleep(tokio::time::Duration::from_millis(200)).await;
            }
        }
        results
    }
}
impl Default for YahooFetcher {
    fn default() -> Self {
        Self::new()
    }
}
#[command]
pub async fn cmd_fetch_yahoo_stocks(symbols: Vec<String>) -> Result<YahooBatchResponse, String> {
    if symbols.is_empty() {
        return Ok(YahooBatchResponse { success: false, message: "No symbols provided".to_string(), total: 0, data: Vec::new() });
    }
    let fetcher = YahooFetcher::new();
    let symbol_refs: Vec<&str> = symbols.iter().map(|s| s.as_str()).collect();
    let stocks = fetcher.fetch_stocks_batch(&symbol_refs).await;
    if stocks.is_empty() {
        Ok(YahooBatchResponse { success: false, message: "Failed to fetch any stock data".to_string(), total: 0, data: Vec::new() })
    } else {
        Ok(YahooBatchResponse { success: true, message: format!("Fetched {} stocks", stocks.len()), total: stocks.len(), data: stocks })
    }
}
